### ISSUE TITLE
`handle` can run when endpoint functions are not exported

### ISSUE THREAD

**Issue by hyunbinseo** [sveltejs/kit#5692]
## Describe the bug

Empty `.js` or `.ts` file that does not export any endpoint functions (eg. `GET`) can still trigger [`handle`](https://kit.svelte.dev/docs/hooks#handle).

## Reproduction

Used `create-svelte@2.0.0-next.149` and `SvelteKit demo app` template.

### Prerequisite

Log request method and pathname in the beginning of the `handle` function.

```ts
export const handle: Handle = async ({ event, resolve }) => {
	const { request: { method }, url: { pathname } } = event;
	console.log(method, pathname);
...
```

### Control Group

Run `vite dev`, open `/`, then navigate to `/about`.

```
  VITE v3.0.2  ready in 205 ms

  ➜  Local:   http://localhost:5174/
  ➜  Network: use --host to expose

GET /
```

### Experimental Group

Create a blank `about.ts` file alongside `about.svelte`

```
├── __layout.svelte
├── about.svelte
├── about.ts // Created
├── index.svelte
└── todos
    ├── _api.ts
    ├── index.svelte
    └── index.ts
```

Run `vite dev`, open `/`, then navigate to `/about`.

```
  VITE v3.0.2  ready in 199 ms

  ➜  Local:   http://localhost:5174/
  ➜  Network: use --host to expose

GET /
GET /about
```

`GET /about` is logged in the experimental group only.

## System Info

```shell
System:
  OS: macOS 12.4
  CPU: (8) arm64 Apple M1
  Memory: 92.70 MB / 8.00 GB
  Shell: 5.8.1 - /bin/zsh
Binaries:
  Node: 16.14.2 - ~/.nvm/versions/node/v16.14.2/bin/node
  npm: 8.5.0 - ~/.nvm/versions/node/v16.14.2/bin/npm
Browsers:
  Chrome: 103.0.5060.134
  Firefox: 102.0
  Safari: 15.5
```

## Severity

annoyance

**Comment by Rich-Harris** [comment]
Hooks run for every request that reaches the server, regardless of whether there's an endpoint involved. This is expected behavior, unless I've misunderstood?

**Comment by hyunbinseo** [comment]
@Rich-Harris Is there a reason why `GET /about` is not logged in the control group?

I was expecting a consistent behavior regardless of an empty `routes/about.ts` file.

**Comment by tcc-sejohnson** [comment]
It's easier to answer why it IS logged in the experimental group first. It's because `about.ts` is a page endpoint for `about.svelte`. When visiting `/about`, `about.svelte` sends a GET request to the server automatically (this request will fail since there's no GET handler implemented, but it has to go through `handle` first).

This should make the lack of logging obvious: When client-side navigating to `/about` when no `about.ts` is present, no GET request for the page endpoint is fired. (There may or may not be a GET request for the page content, but that doesn't cause `handle` to run because it's client-side navigation.) I'd bet if you hard reloaded or navigated to `/about` directly, you'd see a log.

I don't believe there's an issue here.

**Comment by hyunbinseo** [comment]
@tcc-sejohnson I was expecting endpoint requests to be made only when there is a matching request handler function.

Since there is no `GET` request handler exported from the `about.ts` (experimental group), I was expecting a client-side navigation, just like in the control group. (when opening `/` and then navigating to `/about`)

Opening `/about` directly or hard reloading the page does log `GET /about` in both cases.

---

Shouldn't Page Endpoint requests be made only when there is a matching request handler function? (e.g. `GET`)

**Comment by tcc-sejohnson** [comment]
> I was expecting Page Endpoint requests only when there is a matching request handler function.

The client has no idea what functions are exported from the server AFAIK. The existence of a matching `.ts` file determines whether the page gets built with a page endpoint.

> Since there is no GET request handler exported from the about.ts (experimental group), I was expecting a client-side navigation, just like in the control group. (when opening / and then navigating to /about)

You _are_ getting client-side navigation. It's just that your page is making a `fetch` request to its page endpoint for data because you put a matching `.ts` file in there. That `fetch` request is what's triggering `handle`.

> Opening /about directly or hard reloading the page does log GET /about in both cases.

Yep, expected.

TL;DR: Don't create a matching `.ts` file if you don't want your page to try to fetch from its page endpoint.

**Comment by hyunbinseo** [comment]
@tcc-sejohnson Thank you for the replies. I had misconceptions about how a page gets built with a page endpoint.

> The client has no idea what functions are exported from the server AFAIK. The existence of a matching `.ts` file determines whether the page gets built with a page endpoint.

I guess I should not make a `.js` or `.ts` file with a matching name.

I was using it to create a shared request handler for the child routes.

---

I have updated my existing comment from 'client-side navigation' to 'client-side fetch' as you commented.

**Comment by tcc-sejohnson** [comment]
FYI, files starting with an underscore in `routes` do not turn into routes, so that can be a good way to share logic. Also, if you rename the file to `about.json.ts`, it will not become a page endpoint (`about.json !== about`), meaning you can fetch from it: `fetch('/about.json')`.

I mean technically you could name it anything other than literally `about.ts`, but it's convention to name non-page endpoints `.json`, since you're generally hitting them for a json document.

**Comment by hyunbinseo** [comment]
Creating a `.ts` file that starts with an underscore does not generate a `RequestHandler` type.

Since this file is for utility purpose only, I guess I should convert it into a standalone endpoint.

Thank you once again.
